sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala

```diff
@@ -44,20 +52,108 @@ object V2ExpressionUtils extends SQLConfHelper {
+    sequenceToOption(ordering.map(toCatalyst(_, query))).asInstanceOf[Option[Seq[SortOrder]]]
+        .getOrElse(Seq.empty)
+  }
+
+  /**
+   * Converts the V2 [[V2Distribution]] into its counterpart in catalyst [[Distribution]].
+   *
+   * If `funCatalogOpt` is set, it will be used to lookup any V2 function referenced
+   * by the input distribution. For instance, a bucket transform in the distribution requires a
+   * corresponding function to exist in the catalog, in order for Spark to leverage bucket join
+   * for the V2 data sources.
+   *
+   * This returns [[UnspecifiedDistribution]] if any non-identity transform is used in the
+   * distribution, AND the `funCatalogOpt` is not set OR the corresponding function is not
+   * defined in the catalog.
+   */
+  def toCatalystDistribution(
+      distribution: V2Distribution,
+      query: LogicalPlan,
+      funCatalogOpt: Option[FunctionCatalog] = None): Distribution = distribution match {
+    case d: V2OrderedDistribution =>
+      val resolvedOrdering = toCatalystOrdering(d.ordering(), query)
+      OrderedDistribution(resolvedOrdering)
+    case d: V2ClusteredDistribution =>
+      sequenceToOption(d.clustering.map(toCatalyst(_, query, funCatalogOpt)))
+          .map(ClusteredDistribution(_))
+          .getOrElse(UnspecifiedDistribution)
+    case _: V2UnspecifiedDistribution =>
+      UnspecifiedDistribution
+  }
+
+  def toCatalyst(
+      expr: V2Expression,
+      query: LogicalPlan,
+      funCatalogOpt: Option[FunctionCatalog] = None): Option[Expression] = {
     expr match {
+      case t: Transform =>
+        toCatalystTransform(t, query, funCatalogOpt)
       case SortValue(child, direction, nullOrdering) =>
-        val catalystChild = toCatalyst(child, query)
-        SortOrder(catalystChild, toCatalyst(direction), toCatalyst(nullOrdering), Seq.empty)
-      case IdentityTransform(ref) =>
-        resolveRef[NamedExpression](ref, query)
+        toCatalyst(child, query, funCatalogOpt).map { catalystChild =>
+          SortOrder(catalystChild, toCatalyst(direction), toCatalyst(nullOrdering), Seq.empty)
+        }
       case ref: FieldReference =>
-        resolveRef[NamedExpression](ref, query)
+        Some(resolveRef[NamedExpression](ref, query))
       case _ =>
         throw new AnalysisException(s"$expr is not currently supported")
     }
   }
 
+  def toCatalystTransform(
+      trans: Transform,
+      query: LogicalPlan,
+      funCatalogOpt: Option[FunctionCatalog] = None): Option[Expression] = trans match {
+    case IdentityTransform(ref) =>
+      Some(resolveRef[NamedExpression](ref, query))
+    case BucketTransform(numBuckets, refs, sorted) if sorted.isEmpty && refs.length == 1 =>
+      val resolvedRefs = refs.map(r => resolveRef[NamedExpression](r, query))
+      funCatalogOpt.flatMap { catalog =>
+        loadV2Function(catalog, "bucket", resolvedRefs).map { bound =>
+          DataSourceBucketTransformExpression(numBuckets, bound, resolvedRefs)
```

Comment:
How is it different from `DataSourceTransformExpression`? I think Hive bucketed table can also use `DataSourceTransformExpression` in the future.

Reply:
The only difference is it has the `numBuckets` in itself which is used in the equality check. Yes, Hive bucketing can also extend `DataSourceTransformExpression`

Reply:
`numBuckets` can be part of the transform function inputs as an int literal. Seems we only need a singe class `DataSourceTransform`

Reply:
Do you mean we require the V2 `bucket` function to accept two inputs? I think it'll work too.

Reply:
AFAIK the `bucket` transform function support var-length args: `bucket(num_buckets, c1, c2, ...)`

Reply:
sorry I'm confused. What the internal `BoundFunction` means for a bucket function?

Reply:
don't we always check if the inputs are the same first, before checking the function identity? num buckets is a input as well.